### PR TITLE
[io] Add a simple `BufReader`

### PR DESCRIPTION
**Pull request by TheBlueMatt**
Since we require `BufRead` for reading most `rust-bitcoin` objects, its somewhat awkward that we don't provide a `BufReader` in `bitcoin-io`. This fixes that, providing a simple buffered reader for use around `bitcoin_io::Read`s.

It differs from `std`'s largely only in missing functionality and in that it requires a reference to a `Read` rather than a `Read` directly. This makes it more ergonomic as we don't blanket implement `Read for &mut Read` unlike `std`, but also allows for using a `BufReader` without having to take the inner `Read` back out as its always a reference.

This was built clean-room without looking at `std`/`core2`/etc and thus has no license requirements.

### CODE REVIEW

**Comment by tcharding** [comment]
Odd, we have that form all over the place without getting clippy warnings.

**Comment by tcharding** [comment]
I was chatting with @arik-so on signal while he was chatting to you some other way @TheBlueMatt, did you see https://github.com/rust-bitcoin/rust-bitcoin/issues/2766#issuecomment-2284806730 before you hacked this up?

**Review by Kixunil — COMMENTED** [review]
> also allows for using a `BufReader` without having to take the inner `Read` back
out as its always a reference.

What do you mean by this? You should almost never take out a reader from `BufReader` wrapper since that loses buffered data. Usually an application should create the buffered reader right after creating the inner reader and keep it wrapped for the whole lifetime.

Also having a forced reference is not strictly more ergonomic since that prevents usage of owned types. IIRC we originally didn't put blanket impl on `&mut R` because we wanted a blanket impl for `std` types but that was found to be broken so we should be able to put in `&mut R` blanket impl instead. In case we cannot do that we should have a newtype wrapper around references that bypasses the restriction.

File: io/src/lib.rs
Comment (Kixunil) [inline io/src/lib.rs:237]: Maybe the buffer should be const generic.

File: io/src/lib.rs
Comment (Kixunil) [inline io/src/lib.rs:247]: There's some needless zeroing, ideally we would use `ArrayVec<u8>` but sadly we still can't because our read API doesn't support reading into uninitialized buffers (just like std API). We could fix it just like `std` did in nightly but not sure if it's worth it.

File: io/src/lib.rs
Comment (Kixunil) [inline io/src/lib.rs:273]: If `count > self.buf.len()` we should panic instead since it's clearly a bug. (But maybe `debug_assert` is enough.)

File: io/src/lib.rs
Comment (Kixunil) [reply io/src/lib.rs:280]: `debug_assert!(self.pos.saturating_add(amount) < self.buf.len())` would be nice.

File: io/src/lib.rs
Comment (Kixunil) [inline io/src/lib.rs:428]: This makes it look like the reader as whole is limited by the `usize` but it's not. (There's no `self.1 -= len` on purpose.)

The comment should say "returns at most `self.1` number of bytes from each `read` but is otherwise unlimited"

File: io/src/lib.rs
Comment (Kixunil) [reply io/src/lib.rs:274]: I don't think silently coercing the value is helpful. It's most likely harmful since it just produces garbage data. Kinda like UB. Now I also realize that we can either omit the check or turn it into `assert!` because indexing below will panic anyway. I think a nice panic message saying the `Read` implementation is broken would be nicer. (The bound check below will be eliminated by the optimizer in that case.)

File: io/src/lib.rs
Comment (Kixunil) [inline io/src/lib.rs:430]: I think the correct word here is "limited", not "fixed". If the buffer is shorter the number of returned bytes is shorter too.

**Comment by tcharding** [comment]
I added an issue on `rust-bitcoin-maintainer-tools` about the debug assertions thing: https://github.com/rust-bitcoin/rust-bitcoin-maintainer-tools/issues/12

**Comment by tcharding** [comment]
The new `BufReader` doesn't implement `BufRead` (has private inherent methods of the same name as the `BufRead` trait methods)? Can you squash everything down into a single commit also please bro.

File: io/src/lib.rs
Comment (Kixunil) [reply io/src/lib.rs:274]: If we replace `min` with `debug_assert` then `count > self.buf.len()` implies `self.limit > self.buf.len()` which implies panic on `self.buf[..self.limit]`.

Comment (Kixunil) [reply io/src/lib.rs:274]: I'm proposing to remove it. It's just trying to pretend a clearly broken code is not really broken hoping for the best. I don't think we should do that in financial applications.

File: io/src/lib.rs
Comment (Kixunil) [reply io/src/lib.rs:237]: Maybe a generic with default is possible? But if it is, it can also be added later so I'm not going to block the PR over this.

File: io/src/lib.rs
Comment (Kixunil) [inline io/src/lib.rs:271]: Nit: the indexing shouldn't be needed since `&mut [T; N]` coerces to `&mut [T]`.

**Review by Kixunil — APPROVED** [review]
Weak ACK 54360ededcb00cbc18bd5838ebeb2484d03ce573

I still strongly believe that `assert` is better than patching things up with `min` but if other maintainers want to overrule me it won't bother me too much. After all, garbage in - garbage out.

**Review by tcharding — APPROVED** [review]
ACK 54360ededcb00cbc18bd5838ebeb2484d03ce573

**Comment by tcharding** [comment]
I don't have an opinion on the `min` vs `assert` thing and my ACK is in no way meant to override @Kixunil's view.

> Lololol

Glad you were amused by this mistake, writing that review was one of those 'this is too simple, is it _me_ being totally brain dead hear?'

Once again, in-person-meet-ups help lubricate development :)

File: io/src/lib.rs
Comment (Kixunil) [inline io/src/lib.rs:236]: Oh, wait, my questions about the reference weren't addressed. https://github.com/rust-bitcoin/rust-bitcoin/pull/3162#pullrequestreview-2235205845

File: io/src/lib.rs
Comment (apoelstra) [reply io/src/lib.rs]: I think we should use a real `assert`. If we benchmark to find that this is hurting performance we can put the `debug_assert`/`min` combo back.

**Pull request closed without merging**
